pkg/service/backup_routine_handler.go

```diff
@@ -140,64 +159,74 @@ func (h *BackupRoutineHandler) runFullBackupInternal(ctx context.Context, now ti
-			h.storage, h.secretAgent, timebounds, namespace, backupFolder)
+func (h *BackupRoutineHandler) prepareCluster() (*backup.Client, []string, error) {
+	var client *backup.Client
+	var namespaces []string
+	err := h.retry.retry("cluster connection", func() error {
+		var err error
+		client, err = h.clientManager.GetClient(h.backupRoutine.SourceCluster)
 		if err != nil {
-			backupFailureCounter.Inc()
-			return fmt.Errorf("could not start backup of namespace %s, routine %s: %w",
-				namespace, h.routineName, err)
+			return fmt.Errorf("cannot get backup client: %w", err)
+		}
+		namespaces, err = getNamespacesToBackup(h.namespaces, client.AerospikeClient())
+		if err != nil {
+			return fmt.Errorf("cannot retrieve namespaces from source cluster: %w", err)
 		}
 
-		h.fullBackupHandlers[namespace] = handler
-	}
+		return nil
+	})
 
-	return nil
+	return client, namespaces, err
 }
 
-func (h *BackupRoutineHandler) waitForFullBackups(
-	ctx context.Context, backupTimestamp time.Time,
-) error {
-	startTime := time.Now() // startTime is only used to measure backup time
-	for namespace, handler := range h.fullBackupHandlers {
-		backupFolder := getFullPath(h.routineName, h.backupFullPolicy, namespace, backupTimestamp)
-		err := handler.Wait(ctx)
-		if err != nil {
-			backupFailureCounter.Inc()
-			h.logger.Info("Delete failed backup folder",
-				slog.String("path", backupFolder),
+func (h *BackupRoutineHandler) startNamespaceBackup(
+	ctx context.Context, namespace string, now time.Time, client *backup.Client,
+) *RetryableBackupHandler {
+	backupFolder := getFullPath(h.routineName, h.backupFullPolicy, namespace, now)
+	timebounds := h.createTimebounds(now)
+
+	return newRetryableBackupHandler(
+		ctx,
+		h.retry,
+		func(ctx context.Context) (BackupHandler, error) { // start backup.
+			return h.backupService.BackupRun(
+				ctx, h.backupRoutine, h.backupFullPolicy, client, h.storage, h.secretAgent, timebounds, namespace, backupFolder,
 			)
-			h.deleteFolder(ctx, backupFolder) // cleanup on failure
-			return fmt.Errorf("error during backup namespace %s, routine %s: %w",
-				namespace, h.routineName, err)
-		}
+		},
+		func(ctx context.Context) { // on fail.
+			h.deleteFolder(ctx, backupFolder)
+		},
+		func(ctx context.Context, stats *models.BackupStats) error { // on success.
+			return h.writeBackupMetadata(ctx, stats, now, namespace, backupFolder)
+		},
+	)
+}
 
-		if err := h.writeBackupMetadata(ctx, handler.GetStats(), backupTimestamp, namespace, backupFolder); err != nil {
-			return err
+func (h *BackupRoutineHandler) createTimebounds(now time.Time) model.TimeBounds {
+	if h.backupFullPolicy.IsSealed() {
+		return *model.NewTimeBoundsTo(now)
+	}
+	return model.TimeBounds{}
+}
+
+func (h *BackupRoutineHandler) waitForFullBackups(ctx context.Context) error {
+	startTime := time.Now()
+
+	var aggregatedErr error
+	for _, handler := range h.fullBackupHandlers {
+		if err := handler.Wait(ctx); err != nil {
```

Comment:
Is it acceptable for the first handler to potentially block the others due to sequential execution?

Reply:
no, all process are working in separate gorouines